src/Interpreters/ActionsDAG.cpp

```diff
@@ -531,6 +561,115 @@ bool ActionsDAG::empty() const
+    bool ignore_constant_values)
+{
+    size_t num_input_columns = source.size();
+    size_t num_result_columns = result.size();
+
+    if (mode == MatchColumnsMode::Position && num_input_columns != num_result_columns)
+        throw Exception("Number of columns doesn't match", ErrorCodes::NUMBER_OF_COLUMNS_DOESNT_MATCH);
+
+    auto actions_dag = std::make_shared<ActionsDAG>(source);
+    std::vector<Node *> projection(num_result_columns);
+
+    FunctionOverloadResolverPtr func_builder_materialize =
+            std::make_shared<FunctionOverloadResolverAdaptor>(
+                    std::make_unique<DefaultOverloadResolver>(
+                            std::make_shared<FunctionMaterialize>()));
+
+    std::map<std::string_view, std::list<size_t>> inputs;
+    if (mode == MatchColumnsMode::Name)
+    {
+        for (size_t pos = 0; pos < actions_dag->inputs.size(); ++pos)
+            inputs[actions_dag->inputs[pos]->result_name].push_back(pos);
+    }
+
+    for (size_t result_col_num = 0; result_col_num < num_result_columns; ++result_col_num)
+    {
+        const auto & res_elem = result[result_col_num];
+        Node * src_node = nullptr;
+
+        switch (mode)
+        {
+            case MatchColumnsMode::Position:
+            {
+                src_node = actions_dag->inputs[result_col_num];
+                break;
+            }
+
+            case MatchColumnsMode::Name:
+            {
+                auto & input = inputs[res_elem.name];
+                if (input.empty())
+                    throw Exception("Cannot find column " + backQuoteIfNeed(res_elem.name) + " in source stream",
+                                    ErrorCodes::THERE_IS_NO_COLUMN);
+
+                src_node = actions_dag->inputs[input.front()];
+                input.pop_front();
+                break;
+            }
+        }
+
+        /// Check constants.
+
+        if (const auto * res_const = typeid_cast<const ColumnConst *>(res_elem.column.get()))
+        {
+            if (const auto * src_const = typeid_cast<const ColumnConst *>(src_node->column.get()))
+            {
+                if (ignore_constant_values)
+                   src_node = const_cast<Node *>(&actions_dag->addColumn(res_elem, true));
+                else if (res_const->getField() != src_const->getField())
+                    throw Exception("Cannot convert column " + backQuoteIfNeed(res_elem.name) + " because "
+                                    "it is constant but values of constants are different in source and result",
+                                    ErrorCodes::ILLEGAL_COLUMN);
+            }
+            else
+                throw Exception("Cannot convert column " + backQuoteIfNeed(res_elem.name) + " because "
+                                "it is non constant in source stream but must be constant in result",
+                                ErrorCodes::ILLEGAL_COLUMN);
+        }
+
+        if (!res_elem.type->equals(*src_node->result_type))
+        {
```

Comment:
What happens in this scope? I don't understand. Do we simply cast result type to another type?

Reply:
Yes, we just simply apply CAST function
CAST has 2 args - column and new type name
So, we add constant with type name before. And then apply CAST)